pkg/messaging/msgtopic.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Abstract Machines
+// SPDX-License-Identifier: Apache-2.0
+
+package messaging
+
+import (
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/absmach/supermq/pkg/errors"
+)
+
+const (
+	MsgTopicPrefix     = "m"
+	ChannelTopicPrefix = "c"
+
+	numGroups     = 4 // entire expression + domain group + channel group + subtopic group
+	domainGroup   = 1 // domain group is first in msg topic regexp
+	channelGroup  = 2 // channel group is second in msg topic regexp
+	subtopicGroup = 3 // subtopic group is third in msg topic regexp
+)
+
+var (
+	ErrMalformedTopic    = errors.New("malformed topic")
+	ErrMalformedSubtopic = errors.New("malformed subtopic")
+	// Regex to group topic in format m.<domain_id>.c.<channel_id>.<sub_topic> `^\/?m\/([\w\-]+)\/c\/([\w\-]+)(\/[^?]*)?(\?.*)?$`.
+	msgTopicRegExp = regexp.MustCompile(`^\/?` + MsgTopicPrefix + `\/([\w\-]+)\/` + ChannelTopicPrefix + `\/([\w\-]+)(\/[^?]*)?(\?.*)?$`)
+)
+
+func ParsePublishTopic(topic string) (string, string, string, error) {
+	msgParts := msgTopicRegExp.FindStringSubmatch(topic)
+	if len(msgParts) < numGroups {
+		return "", "", "", ErrMalformedTopic
+	}
+
+	domainID := msgParts[domainGroup]
+	chanID := msgParts[channelGroup]
+	subtopic := msgParts[subtopicGroup]
+
+	var err error
+	subtopic, err = ParsePublishSubtopic(subtopic)
+	if err != nil {
+		return "", "", "", errors.Wrap(ErrMalformedTopic, err)
+	}
+
+	return domainID, chanID, subtopic, nil
+}
+
+func ParsePublishSubtopic(subtopic string) (string, error) {
+	if subtopic == "" {
+		return subtopic, nil
+	}
+
+	subtopic, err := url.QueryUnescape(subtopic)
+	if err != nil {
+		return "", errors.Wrap(ErrMalformedSubtopic, err)
+	}
+	subtopic = strings.ReplaceAll(subtopic, "/", ".")
+
+	elems := strings.Split(subtopic, ".")
+	filteredElems := []string{}
+	for _, elem := range elems {
+		if elem == "" {
+			continue
+		}
+
+		if strings.Contains(elem, " ") || strings.Contains(elem, "#") || strings.Contains(elem, "+") || strings.Contains(elem, "*") || strings.Contains(elem, ">") {
```

Comment:
No need to do this in the loop; we can do these checks right before the `replaceAll`:
```go
	for _, r := range allowed {
		if strings.Contains(subtopic, r) {
			return "", ErrMalformedSubtopic
		}
	}
```
where `var allowed = []string{" ", "#", "+", "*", ">"}`.